File: python_cpp/uPIMulator_frontend/src/abi/isa/cc/log_nz_cc.py
from typing import Set

from abi.isa.cc._base_cc import BaseCC
from abi.isa.instruction.condition import Condition


class LogNZCC(BaseCC):
    def __init__(self, condition: Condition):
        super().__init__(
            LogNZCC.conditions(), condition,
        )

    @staticmethod
    def conditions() -> Set[Condition]:
        return {
            Condition.Z,
            Condition.NZ,
            Condition.XZ,
            Condition.XNZ,
            Condition.PL,
            Condition.MI,
            Condition.SZ,
            Condition.SNZ,
            Condition.SPL,
            Condition.SMI,
            Condition.TRUE,
        }

File: golang/uPIMulator/src/linker/kernel/instruction/cc/log_nz_cc.go
package cc

import (
	"errors"
)

type LogNzCc struct {
	condition Condition
}

func (this *LogNzCc) Init(condition Condition) {
	conditions := map[Condition]bool{
		Z:    true,
		NZ:   true,
		XZ:   true,
		XNZ:  true,
		PL:   true,
		MI:   true,
		SZ:   true,
		SNZ:  true,
		SPL:  true,
		SMI:  true,
		TRUE: true,
	}

	if _, found := conditions[condition]; !found {
		err := errors.New("condition is not allowed")
		panic(err)
	}

	this.condition = condition
}

func (this *LogNzCc) Condition() Condition {
	return this.condition
}
